### PR TITLE
Adds support for using Terraform modules cached locally

### LINKED ISSUE #936: Ensure remote modules are downloaded only once

### PR DESCRIPTION
**Pull request by Rchanger**
Add: support for using the terraform init cache for remote modules rather than downloading it again. It will try to use the remote module from terraform cache if --use-terraform-cache flag is set and if the module is not found in cache it will be downloaded

Usage
```
   terrascan scan -d testdata/terraform_cache_use_in_scan -i terraform --use-terraform-cache  
```

Closes #936

### CODE REVIEW

File: pkg/cli/run.go [hunk 1]
```diff
@@ -88,14 +88,17 @@ type ScanOptions struct {
...
 	verbose bool
 
 	// showPassedRules indicates whether to display passed rules or not
 	showPassedRules bool
 
 	// nonRecursive enables recursive scan for the terraform iac provider
 	nonRecursive bool
+
+	// useTerrafomeCache provides ability to use terraform init local cache for modules rather than downloading them.
+	useTerrafomeCache bool
```
Comment (cesar-rodriguez) [inline pkg/cli/run.go:101]: Should this be `useTerraformCache` for consistency?

File: pkg/cli/run.go [hunk 2]
```diff
@@ -181,7 +184,7 @@ func (s *ScanOptions) Run() error {
 
 	// create a new runtime executor for processing IaC
 	executor, err := runtime.NewExecutor(s.iacType, s.iacVersion, s.policyType,
-		s.iacFilePath, s.iacDirPath, s.policyPath, s.scanRules, s.skipRules, s.categories, s.severity, s.nonRecursive)
+		s.iacFilePath, s.iacDirPath, s.policyPath, s.scanRules, s.skipRules, s.categories, s.severity, s.nonRecursive, s.useTerrafomeCache)
```
Comment (cesar-rodriguez) [inline pkg/cli/run.go:187]: Should this be useTerraformCache for consistency?

File: pkg/iac-providers/helm/v3/load-dir.go [hunk 3]
```diff
@@ -42,7 +42,7 @@ var (
 )
 
 // LoadIacDir loads all helm charts under the specified directory
-func (h *HelmV3) LoadIacDir(absRootDir string, nonRecursive bool) (output.AllResourceConfigs, error) {
+func (h *HelmV3) LoadIacDir(absRootDir string, nonRecursive, useTerraformCache bool) (output.AllResourceConfigs, error) {
```
Comment (cesar-rodriguez) [inline pkg/iac-providers/helm/v3/load-dir.go:45]: Why do we need this parameter in IaC providers other than Terraform?

File: pkg/cli/run.go (hunk 1 above)
Comment (Rchanger) [reply pkg/cli/run.go]: yes this was a typo, I have fixed this

**Review by cesar-rodriguez — CHANGES_REQUESTED** [review]
I'm trying to test this feature, but seems like it's not deducting the module directory correctly. I'm using Terraform v0.15.5 with this main.tf file:
```
module "vault" {
  source  = "hashicorp/vault/aws"
  version = "0.16.0"
  ssh_key_name     = ""
  create_dns_entry = false
}
```

When I run:
```
terrascan scan -i terraform --use-terraform-cache -l debug
```

I get this error message and it also downloads the modules instead of using the local cache
```
2021-08-01T22:52:56.103-0400    error   utils/dir.go:49 directory /Users/myusername/programming/vault/.terraform/modules/vault/.terraform/modules does not exist.

2021-08-01T22:52:56.103-0400    debug   downloader/getter.go:49 downloading "github.com/hashicorp/terraform-aws-consul.git?ref=v0.8.0" to "/var/folders/2g/9lkfm6ld2lv350svwr15fdgc0000gn/T/n1ukaa"
```

The cached modules are actually located at:
```
/Users/myusername/programming/vault/.terraform/modules
```

**Review by cesar-rodriguez — DISMISSED** [review]

**Review by cesar-rodriguez — DISMISSED** [review]

**Review by cesar-rodriguez — DISMISSED** [review]

**Review by patilpankaj212 — APPROVED** [review]

File: pkg/iac-providers/terraform/commons/load-dir.go [hunk 4]
```diff
@@ -76,7 +109,7 @@ func (t TerraformDirectoryLoader) LoadIacDir() (allResourcesConfig output.AllRes
 
 	defer t.remoteDownloader.CleanUp()
 
-	if t.nonRecursive {
+	if t.nonRecursive || t.useTerraformCache {
```
Comment (kanchwala-yusuf) [inline pkg/iac-providers/terraform/commons/load-dir.go:112]: Trying to understand the reason for using non recursive scanning with terraform cache?

**Pull request merged**